com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl

```diff
@@ -0,0 +1,657 @@
+    float E_4 = E_3; // Is like this because PQ(L_W)=  1 and PQ(L_B) = 0
+
+    return PQToLinear(E_4, MAX_PQ_VALUE);
+}
+
+
+float3 PerformRangeReduction(float3 input, float minNits, float maxNits)
+{
+    float3 ICtCp = RotateOutputSpaceToICtCp(input); // This is in PQ space.
+    float linearLuma = PQToLinear(ICtCp.x, MAX_PQ_VALUE);
+#if RANGE_REDUCTION == HDRRANGEREDUCTION_REINHARD_LUMA_ONLY
+    linearLuma = ReinhardTonemap(linearLuma, maxNits);
+#elif RANGE_REDUCTION == HDRRANGEREDUCTION_BT2390LUMA_ONLY
+    linearLuma = BT2390EETF(linearLuma, minNits, maxNits);
+#endif
+    ICtCp.x = LinearToPQ(linearLuma);
+
+
+    return RotateICtCpToOutputSpace(ICtCp); // This moves back to linear too!
+}
+
+// TODO: This is very ad-hoc and eyeballed on a limited set. Would be nice to find a standard.
+float3 DesaturateReducedICtCp(float3 ICtCp, float lumaPre, float maxNits)
+{
+    float saturationAmount = min(1.0f, ICtCp.x / max(lumaPre, 1e-6f)); // BT2390, but only when getting darker.
+    //saturationAmount = min(lumaPre / ICtCp.x, ICtCp.x / lumaPre); // Actual BT2390 suggestion
+    saturationAmount *= saturationAmount;
+    //saturationAmount =  pow(smoothstep(1.0f, 0.4f, ICtCp.x), 0.9f);   // A smoothstepp-y function.
+    ICtCp.yz *= saturationAmount;
+    return ICtCp;
+}
+
+float LumaRangeReduction(float input, float minNits, float maxNits, int mode)
+{
+    float3 output = input;
+    if (mode == HDRRANGEREDUCTION_REINHARD)
+    {
+        output = ReinhardTonemap(input, maxNits);
+    }
+    else if (mode == HDRRANGEREDUCTION_BT2390)
+    {
+        output = BT2390EETF(input, minNits, maxNits);
+    }
+
+    return output;
+}
+
+float3 HuePreservingRangeReduction(float3 input, float minNits, float maxNits, int mode)
+{
+    float3 ICtCp = RotateOutputSpaceToICtCp(input);
+
+    float lumaPreRed = ICtCp.x;
+    float linearLuma = PQToLinear(ICtCp.x, MAX_PQ_VALUE);
+    linearLuma = LumaRangeReduction(linearLuma, minNits, maxNits, mode);
+    ICtCp.x = LinearToPQ(linearLuma);
+    ICtCp = DesaturateReducedICtCp(ICtCp, lumaPreRed, maxNits);
+
+    return RotateICtCpToOutputSpace(ICtCp);
+}
+
+float3 HueShiftingRangeReduction(float3 input, float minNits, float maxNits, int mode)
+{
+    float3 hueShiftedResult = input;
+    if (mode == HDRRANGEREDUCTION_REINHARD)
+    {
+        hueShiftedResult.x = ReinhardTonemap(input.x, maxNits);
+        hueShiftedResult.y = ReinhardTonemap(input.y, maxNits);
+        hueShiftedResult.z = ReinhardTonemap(input.z, maxNits);
+    }
+    else if(mode == HDRRANGEREDUCTION_BT2390)
+    {
+        hueShiftedResult.x = BT2390EETF(input.x, minNits, maxNits);
+        hueShiftedResult.y = BT2390EETF(input.y, minNits, maxNits);
+        hueShiftedResult.z = BT2390EETF(input.z, minNits, maxNits);
+    }
+    return hueShiftedResult;
+}
+
+// Ref "High Dynamic Rangecolor grading and displayin Frostbite" [Fry 2017]
+float3 FryHuePreserving(float3 input, float minNits, float maxNits, float hueShift, int mode)
```

Comment:
not used anywhere?

Reply:
No, but I'd like to keep as I expect some feedback at some point about this hue preservation. I rolled our own for the same thing, but if we need to revisit I'd like to have this here.